internal/cb/circuitbreaker.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cb
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	StateOpen   = int32(0)
+	StateClosed = int32(2)
+)
+
+type TryHandler func(ctx context.Context) (interface{}, error)
+type EventHandler func(state int32)
+type RetryPolicyFunc func(currentTry int) time.Duration
+
+type CircuitBreaker struct {
+	// config
+	MaxRetries         int
+	MaxFailureCount    int32
+	ResetTimeout       time.Duration
+	RetryPolicyFunc    RetryPolicyFunc
+	StateChangeHandler EventHandler
+	// state
+	CurrentFailureCount int32
+	State               int32
+}
+
+func NewCircuitBreaker(fs ...CircuitBreakerOptionFunc) *CircuitBreaker {
+	opts, err := NewCircuitBreakerOptions(fs...)
+	if err != nil {
+		panic(fmt.Errorf("evaluating circuitbreaker options: %w", err))
+	}
+	retryPolicyFunc := opts.RetryPolicyFunc
+	if retryPolicyFunc == nil {
+		retryPolicyFunc = func(trial int) time.Duration {
+			return time.Duration(trial) * time.Millisecond
+		}
+	}
+	return &CircuitBreaker{
+		MaxRetries:         opts.MaxRetries,
+		MaxFailureCount:    opts.MaxFailureCount,
+		ResetTimeout:       opts.ResetTimeout,
+		RetryPolicyFunc:    retryPolicyFunc,
+		StateChangeHandler: opts.StateChangeHandler,
+		State:              StateClosed,
+	}
+}
+
+func (cb *CircuitBreaker) Try(tryHandler TryHandler) Future {
+	return cb.TryContext(context.Background(), tryHandler)
+}
+
+func (cb *CircuitBreaker) TryContext(ctx context.Context, tryHandler TryHandler) Future {
+	if state := atomic.LoadInt32(&cb.State); state == StateOpen {
+		return NewFailedFuture(ErrCircuitOpen)
+	}
+	future := NewFutureImpl()
+	go cb.try(ctx, future.resultCh, tryHandler)
+	return future
+}
+
+func (cb *CircuitBreaker) try(ctx context.Context, resultCh chan interface{}, tryHandler TryHandler) {
+	var result interface{}
+	var err error
+loop:
+	for trial := 0; trial <= cb.MaxRetries; trial++ {
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+			break loop
```

Comment:
There should be a place where a client message needs to be converted to error. I traced this errors only to local ones. Are we checking if remove response exception or not anywhere.

Depending on error type of local/remote errors, configuration and type of the request, we should retry it or not. 
I am assuming it is the circuit breakers job, but I could not find the related logic.

To make it more clear, I am talking about this logic.
https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocation.java#L314

Reply:
Error responses are decoded at https://github.com/yuce/hazelcast-go-client/blob/01a76f647faf7b7c7bfa8ca2b793f845b98fc164/internal/cluster/connection.go#L186 and handled at: https://github.com/yuce/hazelcast-go-client/blob/01a76f647faf7b7c7bfa8ca2b793f845b98fc164/internal/invocation/invocation_service.go#L120

circuit breaker retries all errors, I'll fix that.

Reply:
Non-retryable errors implemented at: https://github.com/yuce/hazelcast-go-client/commit/e56053c7635cabdc5b9baffabbc8a8dfe0140e59